lib/spack/spack/cray_manifest.py

```diff
@@ -162,7 +162,8 @@ def entries_to_specs(entries):
                     continue
                 parent_spec = spec_dict[entry["hash"]]
                 dep_spec = spec_dict[dep_hash]
-                parent_spec._add_dependency(dep_spec, deptypes=deptypes)
+                # TODO: Check this with Peter
+                parent_spec._add_dependency(dep_spec, deptypes=deptypes, virtuals=())
```

Comment:
@scheibelp Is :point_up: correct when reading from a Cray manifest? Or are there cases where we read DAGs that should include virtual dependency annotations on edges?

Reply:
Virtual information would not be included in the manifest currently. If `virtuals=()`, does that mean that the dependency provides no virtuals (i.e. `spec["mpi"]` would fail for a dag that contains `openmpi`)? Or does it default to providing all the virtuals that it could provide?

Reply:
It means that the dependency provides no virtuals on the edge being added. Currently I didn't change any type of computation internal to Spack, so `spec["mpi"]` still works without looking at edge attributes - but I'm pretty sure that will change in the future.

When reading from previous formats I have added a function that tries to reconstruct the virtual attributes on edges:

https://github.com/spack/spack/blob/5a41fb2aeccb69a38de18a74df3b86312168d399/lib/spack/spack/spec.py#L4808-L4817

Should we do something similar for Cray too (or reuse the same function if possible)?

Reply:
> It means that the dependency provides no virtuals on the edge being added.

I think we should probably default to inferring that all possible virtuals are provided, but perhaps the next bit will provide clarity on why I think that (and if I'm wrong).

> Should we do something similar for Cray too (or reuse the same function if possible)?

Cray (specifically specs read from manifests on Cray machines) is not different from reading/using specs that were installed prior to this merge. I am confused by the notion that Cray-specific logic is needed:

* If a new install directly depends on a spec read from the manifest, I assume this PR could assign virtual information to the edges
* However, DAGs read from the manifest do not encode the virtual information on the edges, so Specs using these cannot select a subset of virtual functionality (i.e., if `X -> foo` and `foo` provides `blas` and `lapack`, any package that uses `X` must use `foo` for both `blas` and `lapack`)
  * I don't anticipate this will be a problem, as the manifest is primarily for describing things like `cray-libsci` that are direct providers

Does that resolve the issue? Or am I missing something?

Reply:
Done in 9c31dea6ea91ec2493c39f3552715028794256b2 It would be good if someone can suggest how to add a test that checks virtuals are reconstructed properly[^1]

[^1]: I stepped in with a debugger and I believe it works, but it would be good to test it I guess.